### PR TITLE
HIVE-26737: Subquery returning wrong results when database has materialized views

### PR DESCRIPTION
**Pull request by scarlin-cloudera**
When there is a materialized view in the materialized view registry, the HiveMaterializedViewASTSubQueryRewriteShuttle runs and rewrites some RelNodes and RexNodes.

On creation time, the HivePlannerContext is given the RexSubquery nodes which are used to see if it is a correlated subquery with an agg. In the case where the RexSubQuery was rewritten, the Context contains references to stale nodes. It loses the correlated subquery information and creates an incorrect query plan.

The boolean check for subqueries with agg is now done within the Calcite nodes rather than when checking the ASTNodes. It was only used at rule time, so it made more sense for the calculation of the boolean value to be done there, and it's safer in the long run (as opposed to just update the global context when a new RexSubQuery is created).

The HiveFilter and HiveProject will contain the structure holding the calculated correlation information. The information is done with a lazy fetch and only calculated when needed.

The HiveCorrelationInfo structure only contains information for the current subquery level, similar to the old code. A correlated variable cannot go down to a subquery within a subquery at this point.

### What changes were proposed in this pull request?

Specified in description above

### Why are the changes needed?

Specified in description above

### Does this PR introduce _any_ user-facing change?

No

### How was this patch tested?

Added q test which failed before change, succeeds after

### CODE REVIEW

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/correlation/CorrelationInfoVisitor.java [hunk 1]
```diff
@@ -0,0 +1,161 @@
...
+    if (call.getOperator().getKind() == SqlKind.NOT) {
+      notFlag = !notFlag;
+    }
+    RexNode returnNode = super.visitCall(call);
+    // reset flag after visit.
+    if (call.getOperator().getKind() == SqlKind.NOT) {
+      notFlag = !notFlag;
+    }
+    return returnNode;
+  }
```
Comment (kasakrisz) [inline ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/correlation/CorrelationInfoVisitor.java:77]: Could you please add tests when a not expression has multiple subqueries as operands like
```
explain cbo
select id, int_col, year, month from t_test1 s where not (
  s.int_col in (select count(*) from t_test2 t2 where s.id = t2.id) and 
  s.int_col in (select count(*) from t_test3 t3 where s.id = t.3id)
);
```

**Review by kasakrisz — CHANGES_REQUESTED** [review]

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveSubQueryRemoveRule.java [hunk 2]
```diff
@@ -104,44 +106,43 @@ private HiveSubQueryRemoveRule(RelOptRuleOperand operand, String description, Hi
...
       call.transformTo(newRel);
     } else if (relNode instanceof HiveProject) {
-      // if subquery is in PROJECT
       final HiveProject project = call.rel(0);
-      final RexSubQuery e = RexUtil.SubQueryFinder.find(project.getProjects());
-      assert e != null;
+      // Since there is a RexSubQuery, there should be a HiveCorrelationInfo
+      // in the RelNode
+      Preconditions.checkState(project.getCorrelationInfos().size() > 0);
+      HiveCorrelationInfo correlationInfo = project.getCorrelationInfos().get(0);
```
[the remark this answers is not in the document]
Comment (scarlin-cloudera) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveSubQueryRemoveRule.java:133]: You can have two subqueries in 2 different columns which would show up as two different HiveCorrelationInfo objects?

File: ql/src/test/results/clientpositive/perf/tpcds30tb/tez/cbo_query1.q.out [hunk 3]
```diff
@@ -19,8 +19,8 @@ HiveSortLimit(sort0=[$0], dir0=[ASC], fetch=[100])
           HiveProject(s_store_sk=[$0])
             HiveFilter(condition=[=($24, _UTF-16LE'NM')])
               HiveTableScan(table=[[default, store]], table:alias=[store])
-      HiveProject(_o__c0=[*(CAST(/($1, $2)):DECIMAL(21, 6), 1.2:DECIMAL(2, 1))], ctr_store_sk=[$0])
-        HiveFilter(condition=[IS NOT NULL(CAST(/($1, $2)):DECIMAL(21, 6))])
+      HiveProject(ctr_store_sk=[$0], CAST=[CAST(*(CAST(/($1, $2)):DECIMAL(21, 6), 1.2:DECIMAL(2, 1))):DECIMAL(24, 7)])
+        HiveFilter(condition=[IS NOT NULL(CAST(*(CAST(/($1, $2)):DECIMAL(21, 6), 1.2:DECIMAL(2, 1))):DECIMAL(24, 7))])
```
[the remark this answers is not in the document]
Comment (scarlin-cloudera) [reply ql/src/test/results/clientpositive/perf/tpcds30tb/tez/cbo_query1.q.out:23]: Good catch!

Looks like I missed this change, I thought I had only trivial changes.

After looking at this, there was indeed a change of behavior in that it was looking at multiple aggregates in the RelNode stack for "group by" statements when it should only look at the first aggregate.  I made the code change and this regression went away.

File: ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveSubQueryRemoveRule.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (kasakrisz) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveSubQueryRemoveRule.java:132]: I managed to have more than one `CorrelationInfo`:
```
explain cbo
select
  year in (select year from t_test2 where t_test2.id = t_test1.id),
  month in (select month from t_test2 where t_test2.id = t_test1.id)
from t_test1;
```
but here we always get the 1st element of the list and `project.getCorrelationInfos()` is nowhere else is used or I'm missing something.

Could you please check why do we have to collect all `CorrelationInfo` objects

[the remark this answers is not in the document]
Comment (scarlin-cloudera) [reply ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveSubQueryRemoveRule.java:132]: Gonna check this tomorrow, but this is what I think is happening...

I think the rule just happens to get hit twice.  After the first correlation is grabbed, it creates a new filter with one less CorrelationInfo.  The rule gets triggered again and processes the second CorrelationInfo.

On my first pass, I did change this by putting in a for loop.  I had some changes in the plans when I tried that.  And I also figured that this makes the code review cleaner by not adding a for loop?  I can try it again though.  At least if I try it now, the next code review would be only the change of adding a for loop.  Just let me know and I'll try this.

**Review by kasakrisz — APPROVED** [review]

**Pull request merged**